mlir/lib/Transform/AIRDependencyScheduleOpt.cpp

```diff
@@ -2024,6 +2048,37 @@ struct AIRSpecializeChannelWrapAndStrideInScfFor
   }
 
 private:
+  // Clone backward slices of a list of values, opers.
+  SmallVector<Operation *> cloneDefiningOpsInRegion(OpBuilder builder,
+                                                    Region *region,
+                                                    SmallVector<Value> opers,
+                                                    IRMapping &remap) const {
+    SmallVector<Operation *> clonedOps;
+    SetVector<Operation *> backwardSlices;
+    BackwardSliceOptions bsOptions{
+        [&](Operation *o) { return region->isAncestor(o->getParentRegion()); }};
+    if (!region)
+      return clonedOps;
+    for (auto operand : opers) {
+      auto operandDefOp = operand.getDefiningOp();
+      if (!operandDefOp)
+        continue;
+      if (!region->isAncestor(operandDefOp->getParentRegion()))
+        continue;
+      assert(air::isPure(
+          operandDefOp)); // Pure ops are safe to hoist out of loops.
+      // Get backward slices
+      SetVector<Operation *> operandBS;
+      getBackwardSlice(operandDefOp, &operandBS, bsOptions);
+      for (auto b : operandBS)
+        if (air::isPure(b))
+          backwardSlices.insert(b);
```

Comment:
We assert that the operands are pure, but not the operands' operands?

Reply:
Assertion added.